src/commands/mod.rs

```diff
@@ -731,9 +731,9 @@ struct RebaseArgs {
     /// Rebase the whole branch (relative to destination's ancestors)
     #[arg(long, short)]
     branch: Option<RevisionArg>,
-    /// Rebase this revision and its descendants
+    /// Rebase this revision and its descendants (can be repeated)
     #[arg(long, short)]
-    source: Option<RevisionArg>,
+    source: Vec<RevisionArg>,
```

Comment:
Yes, it's not quite obvious what to do here. Whatever we decide, we should probably document it.

The plan in my head was that `jj rebase -r` preserves parent-child relationships, and `jj rebase -s` doesn't. Each commit after `-s` will become a direct child of the destination commit, just as you said in the last paragraph. Certainly, this allows you to do things you couldn't otherwise do. I'm not 100% clear on how useful those things are.

My mental model of `jj rebase -s` is that you cut off a subtree of a tree and carefully move it somewhere else (for each `-s`).

Reply:
I added some docs. PTAL.

Reply:
My 2c: I think it's more consistent (and hence easier to reason about) if `-s X -s Y` is the same as `-s 'X|Y'` (which IIUC, matches @ilyagr's proposed behavior of `-r` -- so not only repeating revisions and using revsets will be consistent but `-r` and `-s` will also be consistent).

> Certainly, this has the advantage of allowing you to do things you couldn't otherwise do. OTOH, I'm not 100% clear on how useful those things are.

It seems to me like it's not all that useful given folks would be hand typing repeated revisions  anyway (unlike revsets), so they can always issue multiple rebase commands if they want `X` and `Y` to be independently rebased (I don't think I'll be using repeating revisions at all but will probably use revsets all the time, so I may be biased).

Reply:
> Certainly, this has the advantage of allowing you to do things you couldn't otherwise do.

Yes, and when `jj rebase -r` eventually supports multiple commits, then you can simply do `jj rebase -r X:` if you want to preserve the shape. So I think what you have is probably good, even if it's different from what `hg rebase -s` does.

@quark-zju: I think I remember that you added support for repeated `-s` and `-b` to `hg rebase`. Do you think it's a mistake to to make `jj rebase -s X:` different from `hg rebase -s X::` by rebasing each commit to be a direct child of the destination commit(s)?

> OTOH, I'm not 100% clear on how useful those things are.

It seems that it would be useful for flattening out a graph, i.e. what I called `jj parallelize` in #1079.

Reply:
>  I think it's more consistent (and hence easier to reason about) if -s X -s Y is the same as -s 'X|Y'

Yes, that is how it works. What gave you the impression it didn't?

> they can always issue multiple rebase commands if they want X and Y to be independently rebased 

This is true

> It seems that it would be useful for flattening out a graph, i.e. what I called jj parallelize in https://github.com/martinvonz/jj/issues/1079.

Interesting point. Unfortunately, this is not something I do often, so I don't have a good sense of it.


-----------------------------------------

I think my primary motivation is that making `jj rebase -s=X` equivalent to `jj rebase '-s=roots(connected(X))'` feels like unnecessary kids gloves. It feels like the distinction can be useful for something (parallelize?), and I'm afraid that if we make these equivalent not, everyone will get used to that and it'll be hard to undo.

We could at first make it an error if `X` is not equal to `roots(connected(X))` and see what usecases come up.

OTOH, we could have it replace `X` with `roots(connected(X))` by default and have an option to disable that (if usecases come up).

As an aside, people typing `-s=X:` accidentally and having each of the commits rebased separately was one of the reasons I felt that guarding it with `--allow-large-revsets` is important.

Reply:
> I think my primary motivation is that making `jj rebase -s=X` equivalent to `jj rebase '-s=roots(connected(X))'` feels like unnecessary kids gloves. It feels like the distinction can be useful for something (parallelize?), and I'm afraid that if we make these equivalent not, everyone will get used to that and it'll be hard to undo.
> 
> We could at first make it an error if `X` is not equal to `roots(connected(X))` and see what usecases come up.
> 
> OTOH, we could have it replace `X` with `roots(connected(X))` by default and have an option to disable that (if usecases come up).

I'd rather keep the behavior you have given it and we can adjust if we get complaints about it.

> As an aside, people typing `-s=X:` accidentally and having each of the commits rebased separately was one of the reasons I felt that guarding it with `--allow-large-revsets` is important.

I think we should rely on undo for that, but I know that undo can be annoying in git-colocated repos. Once we've fixed that, I'm looking forward to removing the `--allow-large-revsets` flag.

Reply:
Sounds great! I'll resolve this then.

> I think we should rely on undo for that, but I know that undo can be annoying in git-colocated repos. ...  I'm looking forward to removing the --allow-large-revsets flag.

Yeah, I know, and I very much agree. This option is obviously a kludge, and `undo` will seem extra magical when it can effortlessly take care of such mistakes.

There are a few minor buts in my mind (What if the UI doesn't make it obvious that something big happened? What about commands that can't be undone, like `jj git push`? What if we hit another undo bug?), but all of them are slightly separate issues that can and should be resolved separately.

Reply:
@avamsi I just realized that my "what gave you the impression that it didn't" question was a bit rude, sorry. I meant it quite literally: I'm wondering if there is some part of the documentation that could be made clearer or added to clarify what's going on. The more eyes we have on the documentation, the better.

Reply:
> @avamsi I just realized that my "what gave you the impression that it didn't" question was a bit rude, sorry.

No worries! FWIW, I didn't think it was rude ("text doesn't convey tone", "assume good intent" etc.).

To answer your question about documentation, I'm not quite sure why I thought they're different, I was probably expecting `-s $REVSET` to preserve shape and so misunderstood your "each -s" remark in one of the replies above (so I thought `-s X|Y` preserves shape but `-s X -s Y` doesn't). Sorry about the noise.

Reply:
> Do you think it's a mistake to to make jj rebase -s X: different from hg rebase -s X:: by rebasing each commit to be a direct child of the destination commit(s)?

I think `hg rebase`'s behavior is better.